cardano-crypto-class/src/Cardano/Crypto/PinnedSizedBytes/TH.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskellQuotes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Crypto.PinnedSizedBytes.TH (
+  psbHex
+  ) where
+
+import Text.Read (readMaybe)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Kind as Haskell
+import Cardano.Crypto.PinnedSizedBytes.Internal (
+  PinnedSizedBytes (PSB)
+  )
+import Data.Primitive.ByteArray (
+  emptyByteArray, 
+  writeByteArray,
+  ByteArray,
+  unsafeFreezeByteArray,
+  newPinnedByteArray,
+  )
+import Control.Monad.ST (runST)
+import Language.Haskell.TH.Quote (
+  QuasiQuoter (
+    QuasiQuoter, 
+    quoteExp,
+    quotePat,
+    quoteType,
+    quoteDec
+    )
+  )
+import Language.Haskell.TH (
+  Q, 
+  Exp (AppTypeE, AppE, ConE, VarE), 
+  Type (LitT),
+  TyLit (NumTyLit),
+  )
+import Control.Monad.Trans.State.Strict (runStateT, StateT, get, modify)
+import Data.Word (Word8)
+import Data.Functor (($>))
+import Data.Foldable (traverse_)
+
+-- | Constructs (and verifies) a compile-time-known hex string as a
+-- 'PinnedSizedBytes', as well as computing its length at the type level.
+--
+-- The syntax accepted by this quasiquoter only works in an expression context.
+-- A valid input consists of the literal sequence @0x@ followed by either:
+--
+-- * A single underscore (@_@); or
```

Comment:
This seems awkward and unnecessary to me.

The only thing that this quasiquoter will ever accept is hex-encoded byte arrays, so there is no need for the leading "0x" (though you could of course make it optional, just to cater for people's motor memory), and the underscore is needlessly complex, both for the user and for the parser. I would propose the following syntax:

- Empty byte array: `[psbHex| |]` or `[psbHex| 0x |]`
- Non-empty byte array: `[psbHex| deadfood |]` or `[psbHex| 0xdeadfood |]`

Reply:
`0x` is technically correct, but my sentiment is that it looks like a typo, which is why I had the underscore to indicate specifically 'I want an empty `PinnedSizedBytes` literal'. I'm happy to have it be `0x` for that meaning.

Reply:
I want to keep the `0x` as it makes copy-pasting literals easier from other sources, such as files or examples. This is necessary for me, as I have outstanding test work for specific SECP-related test vectors.